proposals/Reliable Mutability Primitive.md

```diff
@@ -0,0 +1,166 @@
+
+<!--Explain why this rating-->
+
+3 - As high confidence as we can have without doing actual user studies. We know teams that have tried IPNS but found it unrelaible, ultimately rolling out an alternative solution. We also know teams that have evaluated IPFS but chose alternative due to lack of reliable mutability story.
+
+
+
+### Project definition
+##### Brief plan of attack
+
+<!--Briefly describe the milestones/steps/work needed for this project-->
+
+- Implement third party IPNS republishing (TODO: Find a relevant issue link)
+- Collaborate with Pinning Services to make this part of the API.
+- Fix IPNS publishing in browsers.
+- Enable pubsub by default + IPNS over pubsub.
+- Implement & deploy strategic republishing system on PL hosted nodes that will republish popular content and drop stalled one. So that IPNS works even thought pinning services.
+- (Nice to have) Implement and rollout delegated publishing to enable updating from multiple devices.
+- (Nice to have) Implement and rollout versioning system so that nodes can see full update history.
+
+##### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+Publishing IPNS record takes just a few miliseconds. Shutting laptop and loading `https://ipfs.io/ipns/{key}` week later on other device loads the content published as fast as loading corresponding `https://ipfs.io/ipfs/{cid}`.
+
+
+#####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+<!--
+Provide success criteria. These might include particular metrics, desired changes in the types of bug reports being filed, desired changes in qualitative user feedback (measured via surveys, etc), etc.
+-->
+People and teams are storing IPNS keys in [DNSLink][] records as opposed to CIDs, because that provides a more effective and simpler way to publish updates. We see new projects are leveraging IPNS over working around it via things like [DNSLink][].
+
+
+##### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+- IPNS has bad reputation, changing it may prove to be a challenge.
+- Teams may find that general primitive is not adequate for custom needs.
+- Required key management (e.g. key recovery, key compromise) still may prove it impractical.
+- Higher level alternatives like textile threads may prove to be better layer of abstraction.
+
+##### Alternatives
+_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_
+
+- Integrating textile threads as core component of the ecosystem.
+- Standardizing DNS record updates mechanism and intergating into IPFS.
+- Work with Pinning Services to implement stardized DNSLink updating mechanism.
```

Comment:
@carsonfarmer: [This would be an interesting shorter term solution](https://hackmd.io/@gozala/ipns-pitch)

Reply:
(new person coming up to speed): what would this look like?  I think I'll be able to grok it with a tiny sketch or a mock sample code example.

Reply:
We have recently created [pinning service API](https://ipfs.github.io/pinning-services-api-spec/#section/Schemas) that allows you to add pinning services like pinata to your IPFS node and then remote pin things. Kind of like adding git remote and pushing changes to it.

Idea here is to extend that API in a way that you could update DNSLink provided by the pinning service to the new CID through extended API. Kind of like when you push to remote it updates corresponding branch to new commit. In our case branch names would be DNSLinks, that you could update to point to new CID.

@BigLep does this answer your question ?

Reply:
@Gozala : thanks - this helps.  The only open area for me then is where is this DNSLink stored?  Is this a DNSLink in a DNS record on a (sub)domain that the Pinning Service owns?  

Basically would the chain be
``My domain (e.g., mywebsite.com) has DNS record that points to a domain the Pinning service controls (e.g., mywebsite.somepinningservice.com) which has the DNSLink to the latest CID?``

Reply:
@BigLep Depends if you (A) publish to your own domain, or  (B) use subdomain provided by pinning service for free. 

To give you a real life example, let's say  we want to host https://cid.ipfs.io on some pinning service (pin + take care of updating dnslink + TLS).

(B) is easy. Pinning service simply give you a name on some domain they own. Fleek does that already, they give  https://cid-utils-website.on.fleek.co/  for free + they set up DNSLink for those names (because they own the top domain).

(A) means we want to delegate updates of  DNS TXT record on  `_dnslink.cid.ipfs.io` to some pinning service. 
This can be done either by:
- giving DNS provider credentials to pinning service + expecting them to implement support for proprietary APIs of each dns provider (not realistic IMO, but mentioning here for completeness)
-  using DNS protocol itself for delegation: setting NS records that delegate management of `_dnslink.cid.ipfs.io` to a DNS server controlled by pinning service. That way they have full control over DNSLink without having any control over other records (because we delegated only `_dnslink` subdomain.